src/blockdata/script.rs

```diff
@@ -400,23 +400,31 @@ impl Script {
                                opcodes::All::from(self.0[0]).classify() == opcodes::Class::IllegalOp)
     }
 
-    /// The minimum value an output to a witness script must have in order to be
+    /// The minimum value an output with a P2WPKH script_pubkey must have in order to be
     /// broadcastable on today's bitcoin network.
-    pub const WITNESS_OUTPUT_DUST_THRESHOLD: u64 = 294;
+    pub const P2WPKH_OUTPUT_DUST_THRESHOLD: u64 = 294;
 
-    /// The minimum value an output to a non-witness script must have in order to be
+    /// The minimum value an output with a P2PKH script_pubkey must have in order to be
     /// broadcastable on today's bitcoin network.
-    pub const LEGACY_OUTPUT_DUST_THRESHOLD: u64 = 546;
+    pub const P2PKH_OUTPUT_DUST_THRESHOLD: u64 = 546;
 
     /// Gets the minimum value an output with this script should have in order to be
     /// broadcastable on today's bitcoin network.
     pub fn dust_value(&self) -> u64 {
+        // This must never be lower than Bitcoin Core's GetDustThreshold() (as of v0.21) as it may
+        // otherwise allow users to create transactions which likely can never be
+        // broadcasted/confirmed.
+        3 * // The default dust relay fee is 3000 satoshi/kB (ie 1 sat/vByte)
         if self.is_op_return() {
             0
         } else if self.is_witness_program() {
-            Self::WITNESS_OUTPUT_DUST_THRESHOLD
+            32 + 4 + 1 + (107 / 4) + 4 + // The spend cost copied from Core
+            8 + // The serialized size of the TxOut's amount field
+            self.consensus_encode(&mut ::std::io::sink()).unwrap() as u64 // The serialized size of this script_pubkey
         } else {
-            Self::LEGACY_OUTPUT_DUST_THRESHOLD
+            32 + 4 + 1 + 107 + 4 + // The spend cost copied from Core
+            8 + // The serialized size of the TxOut's amount field
+            self.consensus_encode(&mut ::std::io::sink()).unwrap() as u64 // The serialized size of this script_pubkey
```

Comment:
I will suggest improving readability of this fragment

```rust
       if self.is_op_return() {
             return 0
       }
       let script_pubkey_len =  self.consensus_encode(&mut ::std::io::sink()).expect("I/O sink never fails") as u64;
       let witness_reducer = (self.is_witness_program() as u8) * 4;
       DEFAULT_DUST_RELAY_FEE * // The default dust relay fee is 3000 satoshi/kB (ie 1 sat/vByte)
            (32 + 4 + 1 + 107 / witness_reducer + 4 // The spend cost copied from Core
             + 8  // The serialized size of the TxOut's amount field
             + script_pubkey_len)
```

And introduce `DEFAULT_DUST_RELAY_FEE` const additionally to other values introduced by https://github.com/rust-bitcoin/rust-bitcoin/pull/584

Reply:
I'm not sure that `DUST_RELAY_FEE` is a useful constant - it is only used when calculating the dust threshold, not in policy in any other context. I'm not sure what a user would *do* with such a constant. As for your proposed change, note that `witness_reducer` can be 0 causing a divide-by-0 :).

The current code was written so as to make it clear its directly comparable to Core.